README.md
# Soapy SDR module for SDRPlay

## Documentation

* https://github.com/pothosware/SoapySDRPlay/wiki

## Dependencies

* Get SDR Play driver binaries from - http://sdrplay.com/
* SoapySDR - https://github.com/pothosware/SoapySDR/wiki

## Licensing information

The MIT License (MIT)

Copyright (c) 2015 Charles J. Cliffe

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.



Changelog.txt
Release 0.3.0 (pending)
==========================

- Implement filter by serial number for enumeration

Release 0.2.0 (2019-01-07)
==========================

- support API 2.13 and identify RSPduo

Release 0.1.0 (2016-10-16)
==========================

- Initial release of Soapy SDRPlay support module


Streaming.cpp
/*
 * The MIT License (MIT)
 * 
 * Copyright (c) 2015 Charles J. Cliffe

 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:

 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.

 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include "SoapySDRPlay.hpp"

std::vector<std::string> SoapySDRPlay::getStreamFormats(const int direction, const size_t channel) const 
{
    std::vector<std::string> formats;

    formats.push_back("CS16");
    formats.push_back("CF32");

    return formats;
}

std::string SoapySDRPlay::getNativeStreamFormat(const int direction, const size_t channel, double &fullScale) const 
{
     fullScale = 32767;
     return "CS16";
}

SoapySDR::ArgInfoList SoapySDRPlay::getStreamArgsInfo(const int direction, const size_t channel) const 
{
    SoapySDR::ArgInfoList streamArgs;

    return streamArgs;
}

/*******************************************************************
 * Async thread work
 ******************************************************************/

static void _rx_callback(short *xi, short *xq, unsigned int firstSampleNum, int grChanged, int rfChanged, 
                         int fsChanged, unsigned int numSamples, unsigned int reset, unsigned int hwRemoved, void *cbContext)
{
    SoapySDRPlay *self = (SoapySDRPlay *)cbContext;
    return self->rx_callback(xi, xq, numSamples);
}

static void _gr_callback(unsigned int gRdB, unsigned int lnaGRdB, void *cbContext)
{
    SoapySDRPlay *self = (SoapySDRPlay *)cbContext;
    return self->gr_callback(gRdB, lnaGRdB);
}

void SoapySDRPlay::rx_callback(short *xi, short *xq, unsigned int numSamples)
{
    std::lock_guard<std::mutex> lock(_buf_mutex);

    if (_buf_count == numBuffers)
    {
        _overflowEvent = true;
        return;
    }
    
    int spaceReqd = numSamples * elementsPerSample * shortsPerWord;
    if ((_buffs[_buf_tail].size() + spaceReqd) >= (bufferLength / decM))
    {
       // increment the tail pointer and buffer count
       _buf_tail = (_buf_tail + 1) % numBuffers;
       _buf_count++;

       // notify readStream()
       _buf_cond.notify_one();
    }

    // get current fill buffer
    auto &buff = _buffs[_buf_tail];
    buff.resize(buff.size() + spaceReqd);

    // copy into the buffer queue
    unsigned int i = 0;

    if (useShort)
    {
       short *dptr = buff.data();
       dptr += (buff.size() - spaceReqd);
       for (i = 0; i < numSamples; i++)
       {
           *dptr++ = xi[i];
           *dptr++ = xq[i];
        }
    }
    else
    {
       float *dptr = (float *)buff.data();
       dptr += ((buff.size() - spaceReqd) / shortsPerWord);
       for (i = 0; i < numSamples; i++)
       {
          *dptr++ = (float)xi[i] / 32768.0f;
          *dptr++ = (float)xq[i] / 32768.0f;
       }
    }

    return;
}

void SoapySDRPlay::gr_callback(unsigned int gRdB, unsigned int lnaGRdB)
{
    //Beware, lnaGRdB is really the LNA GR, NOT the LNA state !
 
    mir_sdr_GainValuesT gainVals;

    mir_sdr_GetCurrentGain(&gainVals);

    if (gRdB < 200)
    {
        current_gRdB = gRdB;
    }

    if (gRdB < mir_sdr_GAIN_MESSAGE_START_ID)
    {
        // gainVals.curr is a calibrated gain value
    }
    else if (gRdB == mir_sdr_ADC_OVERLOAD_DETECTED)
    {
        mir_sdr_GainChangeCallbackMessageReceived();
        // OVERLOAD DECTECTED
    }
    else
    {
        mir_sdr_GainChangeCallbackMessageReceived();
        // OVERLOAD CORRECTED
    }
}

/*******************************************************************
 * Stream API
 ******************************************************************/

SoapySDR::Stream *SoapySDRPlay::setupStream(const int direction,
                                            const std::string &format,
                                            const std::vector<size_t> &channels,
                                            const SoapySDR::Kwargs &args)
{
    // check the channel configuration
    if (channels.size() > 1 or (channels.size() > 0 and channels.at(0) != 0)) 
    {
       throw std::runtime_error("setupStream invalid channel selection");
    }
    
    // check the format
    if (format == "CS16") 
    {
        useShort = true;
        shortsPerWord = 1;
        bufferLength = bufferElems * elementsPerSample * shortsPerWord;
        SoapySDR_log(SOAPY_SDR_INFO, "Using format CS16.");
    } 
    else if (format == "CF32") 
    {
        useShort = false;
        shortsPerWord = sizeof(float) / sizeof(short);
        bufferLength = bufferElems * elementsPerSample * shortsPerWord;  // allocate enough space for floats instead of shorts
        SoapySDR_log(SOAPY_SDR_INFO, "Using format CF32.");
    } 
    else 
    {
       throw std::runtime_error( "setupStream invalid format '" + format +
                                  "' -- Only CS16 or CF32 are supported by the SoapySDRPlay module.");
    }

    std::lock_guard<std::mutex> lock(_buf_mutex);

    // clear async fifo counts
    _buf_tail = 0;
    _buf_head = 0;
    _buf_count = 0;

    // allocate buffers
    _buffs.resize(numBuffers);
    for (auto &buff : _buffs) buff.reserve(bufferLength);
    for (auto &buff : _buffs) buff.clear();

    return (SoapySDR::Stream *) this;
}

void SoapySDRPlay::closeStream(SoapySDR::Stream *stream)
{
    std::lock_guard <std::mutex> lock(_general_state_mutex);

    if (streamActive)
    {
        mir_sdr_StreamUninit();
    }
    streamActive = false;
}

size_t SoapySDRPlay::getStreamMTU(SoapySDR::Stream *stream) const
{
    // is a constant in practice
    return bufferElems;
}

int SoapySDRPlay::activateStream(SoapySDR::Stream *stream,
                                 const int flags,
                                 const long long timeNs,
                                 const size_t numElems)
{
    if (flags != 0) 
    {
        return SOAPY_SDR_NOT_SUPPORTED;
    }
   
    resetBuffer = true;
    bufferedElems = 0;
    
    mir_sdr_ErrT err;
    
    std::lock_guard <std::mutex> lock(_general_state_mutex);

    //Enable (= 1) API calls tracing,
    //but only for debug purposes due to its performance impact. 
    mir_sdr_DebugEnable(0);

    //temporary fix for ARM targets.
#if defined(__arm__) || defined(__aarch64__)
    mir_sdr_SetTransferMode(mir_sdr_BULK);
#endif

    err = mir_sdr_StreamInit(&gRdB, sampleRate / 1e6, centerFrequency / 1e6, bwMode,
                             ifMode, lnaState, &gRdBsystem, mir_sdr_USE_RSP_SET_GR, &sps,
                             _rx_callback, _gr_callback, (void *)this);
    if (err != mir_sdr_Success)
    {
       //throw std::runtime_error("StreamInit Error: " + std::to_string(err));
       return SOAPY_SDR_NOT_SUPPORTED;
    }
    mir_sdr_DecimateControl(decEnable, decM, 1);

    mir_sdr_SetDcMode(4,0);
    mir_sdr_SetDcTrackTime(63);
    
    streamActive = true;
    
    return 0;
}

int SoapySDRPlay::deactivateStream(SoapySDR::Stream *stream, const int flags, const long long timeNs)
{
    if (flags != 0)
    {
        return SOAPY_SDR_NOT_SUPPORTED;
    }

    std::lock_guard <std::mutex> lock(_general_state_mutex);

    if (streamActive)
    {
        mir_sdr_StreamUninit();
    }

    streamActive = false;
    
    return 0;
}

int SoapySDRPlay::readStream(SoapySDR::Stream *stream,
                             void * const *buffs,
                             const size_t numElems,
                             int &flags,
                             long long &timeNs,
                             const long timeoutUs)
{   
    if (!streamActive) 
    {
        return 0;
    }
    
    // this is the user's buffer for channel 0
    void *buff0 = buffs[0];
    
    // are elements left in the buffer? if not, do a new read.
    if (bufferedElems == 0)
    {
        int ret = this->acquireReadBuffer(stream, _currentHandle, (const void **)&_currentBuff, flags, timeNs, timeoutUs);
  
        if (ret < 0)
        {
            return ret;
        }
        bufferedElems = ret;
    }

    size_t returnedElems = std::min(bufferedElems.load(), numElems);

    // copy into user's buff0
    if (useShort)
    {
        std::memcpy(buff0, _currentBuff, returnedElems * 2 * sizeof(short));
    }
    else
    {
        std::memcpy(buff0, (float *)_currentBuff, returnedElems * 2 * sizeof(float));
    }
    
    // bump variables for next call into readStream
    bufferedElems -= returnedElems;

    // scope lock here to update _currentBuff position
    {
        std::lock_guard <std::mutex> lock(_buf_mutex);
        _currentBuff += returnedElems * elementsPerSample * shortsPerWord;
    }

    // return number of elements written to buff0
    if (bufferedElems != 0)
    {
        flags |= SOAPY_SDR_MORE_FRAGMENTS;
    }
    else
    {
        this->releaseReadBuffer(stream, _currentHandle);
    }
    return (int)returnedElems;
}

/*******************************************************************
 * Direct buffer access API
 ******************************************************************/

size_t SoapySDRPlay::getNumDirectAccessBuffers(SoapySDR::Stream *stream)
{
    std::lock_guard <std::mutex> lock(_buf_mutex);

    return _buffs.size();
}

int SoapySDRPlay::getDirectAccessBufferAddrs(SoapySDR::Stream *stream, const size_t handle, void **buffs)
{
    std::lock_guard <std::mutex> lock(_buf_mutex);

    buffs[0] = (void *)_buffs[handle].data();
    return 0;
}

int SoapySDRPlay::acquireReadBuffer(SoapySDR::Stream *stream,
                                    size_t &handle,
                                    const void **buffs,
                                    int &flags,
                                    long long &timeNs,
                                    const long timeoutUs)
{
    std::unique_lock <std::mutex> lock(_buf_mutex);

    // reset is issued by various settings
    // overflow set in the rx callback thread
    if (resetBuffer || _overflowEvent)
    {
        // drain all buffers from the fifo
        _buf_tail = 0;
        _buf_head = 0;
        _buf_count = 0;
        for (auto &buff : _buffs) buff.clear();
        _overflowEvent = false;
        if (resetBuffer)
        {
           resetBuffer = false;
        }
        else
        {
           SoapySDR_log(SOAPY_SDR_SSI, "O");
           return SOAPY_SDR_OVERFLOW;
        }
    }

    // wait for a buffer to become available
    if (_buf_count == 0)
    {
        _buf_cond.wait_for(lock, std::chrono::microseconds(timeoutUs));
        if (_buf_count == 0) 
        {
           return SOAPY_SDR_TIMEOUT;
        }
    }

    // extract handle and buffer
    handle = _buf_head;
    buffs[0] = (void *)_buffs[handle].data();
    flags = 0;

    _buf_head = (_buf_head + 1) % numBuffers;

    // return number available
    return (int)(_buffs[handle].size() / (elementsPerSample * shortsPerWord));
}

void SoapySDRPlay::releaseReadBuffer(SoapySDR::Stream *stream, const size_t handle)
{
    std::lock_guard <std::mutex> lock(_buf_mutex);
    _buffs[handle].clear();
    _buf_count--;
}


Registration.cpp
/*
 * The MIT License (MIT)
 * 
 * Copyright (c) 2015 Charles J. Cliffe

 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:

 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.

 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include "SoapySDRPlay.hpp"
#include <SoapySDR/Registry.hpp>

#if !defined(_M_X64) && !defined(_M_IX86)
#define sprintf_s(buffer, buffer_size, stringbuffer, ...) (sprintf(buffer, stringbuffer, __VA_ARGS__))
#endif

static std::map<std::string, SoapySDR::Kwargs> _cachedResults;

static std::vector<SoapySDR::Kwargs> findSDRPlay(const SoapySDR::Kwargs &args)
{
   std::vector<SoapySDR::Kwargs> results;
   unsigned int nDevs = 0;
   char lblstr[128];

   //Enable (= 1) API calls tracing,
   //but only for debug purposes due to its performance impact. 
   mir_sdr_DebugEnable(0);

   std::string baseLabel = "SDRplay Dev";

   // list devices by API
   mir_sdr_DeviceT rspDevs[MAX_RSP_DEVICES];
   mir_sdr_GetDevices(&rspDevs[0], &nDevs, MAX_RSP_DEVICES);

  for (unsigned int i = 0; i < nDevs; i++)
  {
     if (rspDevs[i].devAvail)
     {
        SoapySDR::Kwargs dev;
        dev["serial"] = rspDevs[i].SerNo;
        const bool serialMatch = args.count("serial") == 0 or args.at("serial") == dev["serial"];
        if (not serialMatch) continue;
        if (rspDevs[i].hwVer > 253)
        {
           sprintf_s(lblstr, sizeof(lblstr), "SDRplay Dev%d RSP1A %s", i, rspDevs[i].SerNo);
        }
        else if (rspDevs[i].hwVer == 3)
        {
           sprintf_s(lblstr, sizeof(lblstr), "SDRplay Dev%d RSPduo %s", i, rspDevs[i].SerNo);
        }
        else
        {
           sprintf_s(lblstr, sizeof(lblstr), "SDRplay Dev%d RSP%d %s", i, rspDevs[i].hwVer, rspDevs[i].SerNo);
        }
        dev["label"] = lblstr;
        results.push_back(dev);
        _cachedResults[rspDevs[i].SerNo] = dev;
     }
  }

    //fill in the cached results for claimed handles
    for (const auto &serial : SoapySDRPlay_getClaimedSerials())
    {
        if (_cachedResults.count(serial) == 0) continue;
        if (args.count("serial") != 0 and args.at("serial") != serial) continue;
        results.push_back(_cachedResults.at(serial));
    }

   return results;
}

static SoapySDR::Device *makeSDRPlay(const SoapySDR::Kwargs &args)
{
    return new SoapySDRPlay(args);
}

static SoapySDR::Registry registerSDRPlay("sdrplay", &findSDRPlay, &makeSDRPlay, SOAPY_SDR_ABI_VERSION);


SoapySDRPlay.hpp
/*
 * The MIT License (MIT)
 * 
 * Copyright (c) 2015 Charles J. Cliffe

 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:

 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.

 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
#pragma once

#include <SoapySDR/Device.hpp>
#include <SoapySDR/Logger.h>
#include <SoapySDR/Types.h>
#include <stdexcept>
#include <thread>
#include <mutex>
#include <atomic>
#include <condition_variable>
#include <string>
#include <cstring>
#include <algorithm>
#include <set>

#ifdef _WIN32
#include <mir_sdr.h>
#else
#include <mirsdrapi-rsp.h>
#endif

#define DEFAULT_BUFFER_LENGTH     (65536)
#define DEFAULT_NUM_BUFFERS       (8)
#define DEFAULT_ELEMS_PER_SAMPLE  (2)

#define MAX_RSP_DEVICES  (4)

std::set<std::string> &SoapySDRPlay_getClaimedSerials(void);

class SoapySDRPlay: public SoapySDR::Device
{
public:
    SoapySDRPlay(const SoapySDR::Kwargs &args);

    ~SoapySDRPlay(void);

    /*******************************************************************
     * Identification API
     ******************************************************************/

    std::string getDriverKey(void) const;

    std::string getHardwareKey(void) const;

    SoapySDR::Kwargs getHardwareInfo(void) const;

    /*******************************************************************
     * Channels API
     ******************************************************************/

    size_t getNumChannels(const int) const;

    /*******************************************************************
     * Stream API
     ******************************************************************/

    std::vector<std::string> getStreamFormats(const int direction, const size_t channel) const;

    std::string getNativeStreamFormat(const int direction, const size_t channel, double &fullScale) const;

    SoapySDR::ArgInfoList getStreamArgsInfo(const int direction, const size_t channel) const;

    SoapySDR::Stream *setupStream(const int direction, 
                                  const std::string &format, 
                                  const std::vector<size_t> &channels = std::vector<size_t>(), 
                                  const SoapySDR::Kwargs &args = SoapySDR::Kwargs());

    void closeStream(SoapySDR::Stream *stream);

    size_t getStreamMTU(SoapySDR::Stream *stream) const;

    int activateStream(SoapySDR::Stream *stream,
                       const int flags = 0,
                       const long long timeNs = 0,
                       const size_t numElems = 0);

    int deactivateStream(SoapySDR::Stream *stream, const int flags = 0, const long long timeNs = 0);

    int readStream(SoapySDR::Stream *stream,
                   void * const *buffs,
                   const size_t numElems,
                   int &flags,
                   long long &timeNs,
                   const long timeoutUs = 200000);

    /*******************************************************************
     * Direct buffer access API
     ******************************************************************/

    size_t getNumDirectAccessBuffers(SoapySDR::Stream *stream);

    int getDirectAccessBufferAddrs(SoapySDR::Stream *stream, const size_t handle, void **buffs);

    int acquireReadBuffer(SoapySDR::Stream *stream,
                          size_t &handle,
                          const void **buffs,
                          int &flags,
                          long long &timeNs,
                          const long timeoutUs = 100000);

    void releaseReadBuffer(SoapySDR::Stream *stream, const size_t handle);

    /*******************************************************************
     * Antenna API
     ******************************************************************/

    std::vector<std::string> listAntennas(const int direction, const size_t channel) const;

    void setAntenna(const int direction, const size_t channel, const std::string &name);

    std::string getAntenna(const int direction, const size_t channel) const;

    /*******************************************************************
     * Frontend corrections API
     ******************************************************************/

    bool hasDCOffsetMode(const int direction, const size_t channel) const;

    /*******************************************************************
     * Gain API
     ******************************************************************/

    std::vector<std::string> listGains(const int direction, const size_t channel) const;

    bool hasGainMode(const int direction, const size_t channel) const;

    void setGainMode(const int direction, const size_t channel, const bool automatic);

    bool getGainMode(const int direction, const size_t channel) const;

    void setGain(const int direction, const size_t channel, const std::string &name, const double value);

    double getGain(const int direction, const size_t channel, const std::string &name) const;

    SoapySDR::Range getGainRange(const int direction, const size_t channel, const std::string &name) const;

    /*******************************************************************
     * Frequency API
     ******************************************************************/

    void setFrequency(const int direction,
                      const size_t channel,
                      const std::string &name,
                      const double frequency,
                      const SoapySDR::Kwargs &args = SoapySDR::Kwargs());

    double getFrequency(const int direction, const size_t channel, const std::string &name) const;

    SoapySDR::RangeList getBandwidthRange(const int direction, const size_t channel) const;
    
    std::vector<std::string> listFrequencies(const int direction, const size_t channel) const;

    SoapySDR::RangeList getFrequencyRange(const int direction, const size_t channel, const std::string &name) const;

    SoapySDR::ArgInfoList getFrequencyArgsInfo(const int direction, const size_t channel) const;

    /*******************************************************************
     * Sample Rate API
     ******************************************************************/

    void setSampleRate(const int direction, const size_t channel, const double rate);

    double getSampleRate(const int direction, const size_t channel) const;

    std::vector<double> listSampleRates(const int direction, const size_t channel) const;

    /*******************************************************************
    * Bandwidth API
    ******************************************************************/

    void setBandwidth(const int direction, const size_t channel, const double bw);

    double getBandwidth(const int direction, const size_t channel) const;

    std::vector<double> listBandwidths(const int direction, const size_t channel) const;
    
    void setDCOffsetMode(const int direction, const size_t channel, const bool automatic);
    
    bool getDCOffsetMode(const int direction, const size_t channel) const;
    
    bool hasDCOffset(const int direction, const size_t channel) const;

    /*******************************************************************
     * Settings API
     ******************************************************************/

    SoapySDR::ArgInfoList getSettingInfo(void) const;

    void writeSetting(const std::string &key, const std::string &value);

    std::string readSetting(const std::string &key) const;

    /*******************************************************************
     * Async API
     ******************************************************************/

    void rx_callback(short *xi, short *xq, unsigned int numSamples);

    void gr_callback(unsigned int gRdB, unsigned int lnaGRdB);

private:

    /*******************************************************************
     * Internal functions
     ******************************************************************/

    static double getRateForBwEnum(mir_sdr_Bw_MHzT bwEnum);

    static uint32_t getInputSampleRateAndDecimation(uint32_t rate, unsigned int *decM, unsigned int *decEnable, mir_sdr_If_kHzT ifMode);

    static mir_sdr_Bw_MHzT getBwEnumForRate(double rate, mir_sdr_If_kHzT ifMode);

    static  double getBwValueFromEnum(mir_sdr_Bw_MHzT bwEnum);

    static mir_sdr_Bw_MHzT mirGetBwMhzEnum(double bw);

    static mir_sdr_If_kHzT stringToIF(std::string ifMode);

    static std::string IFtoString(mir_sdr_If_kHzT ifkHzT);

    /*******************************************************************
     * Private variables
     ******************************************************************/
    //device settings
    mir_sdr_Bw_MHzT bwMode;
    mir_sdr_If_kHzT ifMode;
    float ver;

    int gRdB;
    std::atomic_int current_gRdB;
    int gRdBsystem;
    int sps;
    int lnaState;
    int hwVer;

    //cached settings
    uint32_t sampleRate;
    uint32_t reqSampleRate;
    unsigned int decM;
    unsigned int decEnable;
    uint32_t centerFrequency;
    double ppm;
    std::atomic_int bufferLength;

    //numBuffers, bufferElems, elementsPerSample
    //are indeed constants
    const size_t numBuffers = DEFAULT_NUM_BUFFERS;
    const unsigned int bufferElems = DEFAULT_BUFFER_LENGTH;
    const int elementsPerSample = DEFAULT_ELEMS_PER_SAMPLE;

    std::atomic_uint shortsPerWord;
 
    mir_sdr_AgcControlT agcMode;
    std::atomic_bool streamActive;
  
    bool dcOffsetMode;
    std::atomic_bool useShort;

    unsigned int IQcorr;
    int setPoint;

    mir_sdr_RSPII_AntennaSelectT antSel;
    mir_sdr_rspDuo_TunerSelT tunSel;
    int amPort;
    unsigned int extRef;
    unsigned int biasTen;
    unsigned int notchEn;
    unsigned int dabNotchEn;
    std::string serNo;

public:

   /*******************************************************************
    * Public variables
    ******************************************************************/
    
    mutable std::mutex _general_state_mutex;

    std::mutex _buf_mutex;
    std::condition_variable _buf_cond;

    std::vector<std::vector<short> > _buffs;
    size_t	_buf_head;
    size_t	_buf_tail;
    size_t	_buf_count;
    short *_currentBuff;
    bool _overflowEvent;
    std::atomic_size_t bufferedElems;
    size_t _currentHandle;
    std::atomic_bool resetBuffer;
};


Registation.cpp
/*
 * The MIT License (MIT)
 * 
 * Copyright (c) 2015 Charles J. Cliffe

 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:

 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.

 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include "SoapySDRPlay.hpp"
#include <SoapySDR/Registry.hpp>

#if !defined(_M_X64) && !defined(_M_IX86)
#define sprintf_s(buffer, buffer_size, stringbuffer, ...) (sprintf(buffer, stringbuffer, __VA_ARGS__))
#endif

#define MAX_RSP_DEVICES  (4)

static mir_sdr_DeviceT rspDevs[MAX_RSP_DEVICES];
bool deviceSelected = false;

static std::vector<SoapySDR::Kwargs> findSDRPlay(const SoapySDR::Kwargs &args)
{
   std::vector<SoapySDR::Kwargs> results;
   std::string labelHint;
   if (args.count("label") != 0) labelHint = args.at("label");
   unsigned int nDevs = 0;
   char lblstr[128];

   if (deviceSelected == true)
   {
      mir_sdr_ReleaseDeviceIdx();
      deviceSelected = false;
   }
   //Enable (= 1) API calls tracing,
   //but only for debug purposes due to its performance impact. 
   mir_sdr_DebugEnable(0);

   std::string baseLabel = "SDRplay Dev";

	// list devices by API
   mir_sdr_GetDevices(&rspDevs[0], &nDevs, MAX_RSP_DEVICES);

   size_t posidx = labelHint.find(baseLabel);

   if (posidx != std::string::npos)
   {
      unsigned int devIdx = labelHint.at(posidx + baseLabel.length()) - 0x30;

      if ((devIdx < nDevs) && (rspDevs[devIdx].devAvail))
      {
         SoapySDR::Kwargs dev;
         dev["driver"] = "sdrplay";
         if (rspDevs[devIdx].hwVer > 253)
         {
             sprintf_s(lblstr, 128, "SDRplay Dev%d RSP1A %d %s", devIdx, rspDevs[devIdx].hwVer, rspDevs[devIdx].SerNo);
         }
         else
         {
             sprintf_s(lblstr, 128, "SDRplay Dev%d RSP%d %s", devIdx, rspDevs[devIdx].hwVer, rspDevs[devIdx].SerNo);
         }
         dev["label"] = lblstr;
         results.push_back(dev);
      }
   }
   else
   {
      for (unsigned int i = 0; i < nDevs; i++)
      {
         if (rspDevs[i].devAvail)
         {
            SoapySDR::Kwargs dev;
            dev["driver"] = "sdrplay";
            if (rspDevs[i].hwVer > 253)
            {
               sprintf_s(lblstr, 128, "SDRplay Dev%d RSP1A %d %s", i, rspDevs[i].hwVer, rspDevs[i].SerNo);
            }
            else
            {
               sprintf_s(lblstr, 128, "SDRplay Dev%d RSP%d %s", i, rspDevs[i].hwVer, rspDevs[i].SerNo);
            }
            dev["label"] = lblstr;
            results.push_back(dev);
         }
      }
   }
   return results;
}

static SoapySDR::Device *makeSDRPlay(const SoapySDR::Kwargs &args)
{
    return new SoapySDRPlay(args);
}

static SoapySDR::Registry registerSDRPlay("sdrplay", &findSDRPlay, &makeSDRPlay, SOAPY_SDR_ABI_VERSION);
